pkg/controller/machineset/controller.go

```diff
@@ -68,61 +87,43 @@ func (c *MachineSetControllerImpl) Reconcile(machineSet *v1alpha1.MachineSet) er
 		return err
 	}
 
-	return c.syncReplicas(machineSet, filteredMachines)
-}
-
-func (c *MachineSetControllerImpl) Get(namespace, name string) (*v1alpha1.MachineSet, error) {
-	return c.machineSetsLister.MachineSets(namespace).Get(name)
-}
-
-// syncReplicas essentially scales machine resources up and down.
-func (c *MachineSetControllerImpl) syncReplicas(machineSet *v1alpha1.MachineSet, machines []*v1alpha1.Machine) error {
 	// Take ownership of machines if not already owned.
-	for _, machine := range machines {
+	for _, machine := range filteredMachines {
 		if shouldAdopt(machineSet, machine) {
 			c.adoptOrphan(machineSet, machine)
 		}
 	}
 
-	var result error
-	currentMachineCount := int32(len(machines))
-	desiredReplicas := *machineSet.Spec.Replicas
-	diff := int(currentMachineCount - desiredReplicas)
+	var manageReplicasErr error
+	if machineSet.DeletionTimestamp == nil {
+		manageReplicasErr = c.manageReplicas(filteredMachines, machineSet)
+	}
+	ms := machineSet.DeepCopy()
+	newStatus := c.calculateStatus(ms, filteredMachines)
 
-	if diff < 0 {
-		diff *= -1
-		for i := 0; i < diff; i++ {
-			glog.V(2).Infof("creating a machine ( spec.replicas(%d) > currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
-			machine, err := c.createMachine(machineSet)
-			if err != nil {
-				return err
-			}
-			_, err = c.machineClient.ClusterV1alpha1().Machines(machineSet.Namespace).Create(machine)
-			if err != nil {
-				glog.Errorf("unable to create a machine = %s, due to %v", machine.Name, err)
-				result = err
-			}
-		}
-	} else if diff > 0 {
-		for i := 0; i < diff; i++ {
-			glog.V(2).Infof("deleting a machine ( spec.replicas(%d) < currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
-			// TODO: Define machines deletion policies.
-			// see: https://github.com/kubernetes/kube-deploy/issues/625
-			machineToDelete := machines[i]
-			err := c.machineClient.ClusterV1alpha1().Machines(machineSet.Namespace).Delete(machineToDelete.Name, &metav1.DeleteOptions{})
-			if err != nil {
-				glog.Errorf("unable to delete a machine = %s, due to %v", machineToDelete.Name, err)
-				result = err
-			}
-		}
+	// Always updates status as machines come up or die.
+	updatedMS, err := updateMachineSetStatus(c.clusterAPIClient.ClusterV1alpha1().MachineSets(machineSet.Namespace), machineSet, newStatus)
+	if err != nil {
+		// Multiple things could lead to this update failing. Requeuing the replica set ensures
+		// Returning an error causes a requeue without forcing a hotloop
+		return err
 	}
+	// Resync the MachineSet after MinReadySeconds as a last line of defense to guard against clock-skew.
+	if manageReplicasErr == nil && updatedMS.Spec.MinReadySeconds > 0 &&
+		updatedMS.Status.ReadyReplicas == *(updatedMS.Spec.Replicas) &&
```

Comment:
Need more commentary to understand this scenario. I do not understand the scenario where all replicas are ready but some are not available. Shouldn't a replica be available to be counted as ready?

Reply:
added more comments.

	// Clock-skew is an issue as it may impact whether a ready replica is counted as an available replica.
	// A replica is available if the amount of time since last transition exceeds MinReadySeconds.
	// If there was a clock skew, checking whether the amount of time since last transition to ready state
	// exceeds MinReadySeconds could be incorrect.
	// To avoid an available replica stuck in the ready state, we force a reconcile after MinReadySeconds,
	// at which point it should confirm any available replica to be available.